jmeter/src/main/java/org/eclipse/hono/jmeter/client/HonoCommander.java

```diff
@@ -0,0 +1,252 @@
+        final CompletableFuture<Void> shutdownTracker = new CompletableFuture<>();
+        final Future<Void> clientTracker = Future.future();
+        LOG.trace("Clean resources...");
+        client.shutdown(clientTracker.completer());
+        clientTracker
+                .compose(ok -> closeVertx())
+                .recover(error -> closeVertx())
+                .setHandler(result -> shutdownTracker.complete(null));
+        return shutdownTracker;
+    }
+
+    private Future<MessageConsumer> createConsumers() {
+        return client
+                .createEventConsumer(sampler.getTenant(),
+                        getConsumer(message -> handleMessage("Event", message),
+                                this::handleCommandReadinessNotification),
+                        closeHook -> LOG.error("remotely detached consumer link"))
+                .compose(consumer -> client.createTelemetryConsumer(sampler.getTenant(),
+                        getConsumer(message -> handleMessage("Telemetry", message),
+                                this::handleCommandReadinessNotification),
+                        closeHook -> LOG.error("remotely detached consumer link")))
+                .map(consumer -> {
+                    LOG.info("Ready to receive command readiness notifications for tenant [{}]",
+                            sampler.getTenant());
+                    return consumer;
+                })
+                .recover(Future::failedFuture);
+    }
+
+    private void handleMessage(final String msgType, final Message msg) {
+        final Data body = (Data) msg.getBody();
+        LOG.trace("Type: [{}] and Message: [{}]", msgType, body != null ? body.getValue().toString() : "");
+    }
+
+    private void handleCommandReadinessNotification(final TimeUntilDisconnectNotification notification) {
+        synchronized (lock) {
+            if (notification.getMillisecondsUntilExpiry() == 0) {
+                LOG.trace("Device [{}:{}] not ready to receive commands.", notification.getTenantId(),
+                        notification.getDeviceId());
+                if (!"http".equals(sampler.getProtocol())) {
+                    allConnectedDevices.remove(notification.getDeviceId());
+                }
+            } else {
+                LOG.trace("Sending command to device [{}:{}]", notification.getTenantId(), notification.getDeviceId());
+                if ("http".equals(sampler.getProtocol())) {
+                    sendCommandAndReceiveResponse(notification.getTenantId(), notification.getDeviceId());
+                } else if ("mqtt".equals(sampler.getProtocol())) {
+                    LOG.trace("A device [{}] got registered over mqtt", notification.getDeviceId());
+                    allConnectedDevices.add(notification.getDeviceId());
+                }
+            }
+        }
+    }
+
+    private void sendCommandAndReceiveResponse(final String tenantId, final String deviceId) {
+        synchronized (lock) {
```

Comment:
I don't think the synchronized does what you had in mind here. The following calls build up a chain of calls, but doesn't actually execute them. So e.g. the call to `client.getOrCreateCommandClient` will hold the lock, but the lambda of the following `compose`might not.

Reply:
Thank you for the input. My intention was not to ensure the execution of vertx future, rather to ensure that only one thread operates on this at a time. Inside the handler, there is a piece of code to increment success or error counts (which are field variables) and the idea is to ensure that this part of code is handled by only one thread at a time. Ofcourse only this particular part can be placed inside synchronised.

As part of the other changes, currently I am replacing this part of code to use vertx eventbus to notify the success or the failure. You can have have a look once I pushed the code.

Reply:
Pls ignore my last comment, I interpreted your comment differently. :) I understood now what you meant in your last comment with lock and compose. In the last push, I synchronized the block inside the mapper instead of doing it as before.